### ISSUE TITLE
Gradient Ascent Cross Entropy Loss

### ISSUE THREAD

**Issue by seanremy** [pytorch/pytorch#3495]
I'm using a GAN-like setup using CrossEntropyLoss and am curious about the best way to do gradient ascent. Since the one-hot conversion takes place inside the loss function, I am just reversing the gradients as follows:

```
loss = criterion(outputs, labels)
loss.backward()
for group in optimizer.param_groups:
	for p in group['params']:
		p.grad = -1*p.grad
optimizer.step()
```

Is this the correct way to go about this? My model outputs quickly become nan when I do this, and I'm trying to figure out if the above code is the reason.

**Comment by SsnL** [comment]
Your code should just work. The problem lies elsewhere. However, why don't you use `(-loss).backward()`? Is it because there is an adversary needing the `loss.backward()` gradient?

Please post such questions to the forum discuss.pytorch.org, rather than as a github issue.

**Comment by seanremy** [comment]
Okay, that's good to know. And that's exactly why.

Thanks, had no idea that forum was even there, I'll use that from now on!

**Comment by SsnL** [comment]
I'd try reducing the lr, or just removing the updates for one of your network (the one your want or the adversarial one), and see where the problem is.

**Comment by seanremy** [comment]
So I've tried reducing the lr but it has to get down to about 1e-10 to prevent the loss from increasing, and then it is too small for the model to get anywhere.

**Comment by seanremy** [comment]
And the discriminator is set to not update by setting all of its layers .trainable to False

**Comment by cdjhz** [comment]
> Your code should just work. The problem lies elsewhere. However, why don't you use `(-loss).backward()`? Is it because there is an adversary needing the `loss.backward()` gradient?
> 
> Please post such questions to the forum discuss.pytorch.org, rather than as a github issue.

Hey! is there a way to implement gradient ascent in pytorch? In some cases, some of the terms in the loss are maximized for one network and minimized for another network. But, it seems the learning rate must be set positive. Otherwise, I have to to forward the data again and compute the -loss and backward it.

**Comment by SsnL** [comment]
@cdjhz Alternatively write a custom `torch.optim.Optimizer`

**Comment by pooriakh** [comment]
> I'm using a GAN-like setup using CrossEntropyLoss and am curious about the best way to do gradient ascent. Since the one-hot conversion takes place inside the loss function, I am just reversing the gradients as follows:
> 
> ```
> loss = criterion(outputs, labels)
> loss.backward()
> for group in optimizer.param_groups:
> 	for p in group['params']:
> 		p.grad = -1*p.grad
> optimizer.step()
> ```
> 
> Is this the correct way to go about this? My model outputs quickly become nan when I do this, and I'm trying to figure out if the above code is the reason.


Thanks, but I would like to complete your code snippet by adding a single line because sometimes there will be None in the gradients

                for group in optimizer.param_groups:
                    for p in group['params']:
                        if p.grad is not None:
                            p.grad = -1*p.grad

**Comment by pooriakh** [comment]
add a single line for removing errors caused by the occurrence of None
